hypothesis-python/src/hypothesis/provisional.py

```diff
@@ -89,34 +89,45 @@ def __init__(
         # information in https://tools.ietf.org/html/rfc1035#section-2.3.1
         # which defines the allowed syntax of a subdomain string.
         if self.max_element_length == 1:
-            self.label_regex = r"[a-zA-Z]"
+            label_regex = r"[a-zA-Z]"
         elif self.max_element_length == 2:
-            self.label_regex = r"[a-zA-Z][a-zA-Z0-9]?"
+            label_regex = r"[a-zA-Z][a-zA-Z0-9]?"
         else:
             maximum_center_character_pattern_repetitions = self.max_element_length - 2
-            self.label_regex = r"[a-zA-Z]([a-zA-Z0-9\-]{0,%d}[a-zA-Z0-9])?" % (
+            label_regex = r"[a-zA-Z]([a-zA-Z0-9\-]{0,%d}[a-zA-Z0-9])?" % (
                 maximum_center_character_pattern_repetitions,
             )
 
-    def do_draw(self, data):
+        # Construct reusable strategies here to avoid a performance hit by doing
+        # so repeatedly in do_draw.
+
         # 1 - Select a valid top-level domain (TLD) name
         # 2 - Check that the number of characters in our selected TLD won't
         # prevent us from generating at least a 1 character subdomain.
         # 3 - Randomize the TLD between upper and lower case characters.
-        domain = data.draw(
+        @st.composite
+        def recase_randomly(draw, tld):
+            tld = list(tld)
+            changes = draw(st.tuples(*(st.booleans() for _ in range(len(tld)))))
+            for i, change_case in enumerate(changes):
+                if change_case:
+                    tld[i] = tld[i].lower() if tld[i].isupper() else tld[i].upper()
+            return "".join(tld)
+
+        self.domain_strategy = (
             st.sampled_from(get_top_level_domains())
```

Comment:
The previous flatmap dynamically created a lot of uncacheable strategies. variable-length tuples aren't enormously cacheable in general, but they are decently so when the length is the only thing that can vary.
```
            .flatmap(
                lambda tld: st.tuples(
                    *(st.sampled_from([c.lower(), c.upper()]) for c in tld)
                ).map("".join)
            )
```